base/inc/AdePT/TrackManager.h

```diff
@@ -0,0 +1,263 @@
+  adept::MParray *fNextTracks{nullptr};   ///< Array of rack slots for the next iteration (device pointer)
+  Track *fBuffer{nullptr};                ///< Storage for the circular buffer of tracks (device pointer)
+
+  /// @brief Construction done on host but holding device pointers.
+  __host__ __device__ TrackManager(size_t capacity) : fCapacity(capacity)
+  {
+    // The Track type must be copy constructible
+    if (!std::is_copy_constructible<Track>::value)
+      COPCORE_EXCEPTION("TrackManager ctor: The track type must be copy constructible");
+    // The queue size must be a power of 2 (for fast access in the circular buffer)
+    bool pow_two = (capacity & (capacity - 1)) == 0;
+    if (!pow_two) COPCORE_EXCEPTION("TrackManager ctor: The buffer size has to be a power of 2");
+    fNextFree.store(0);
+  }
+
+  /// Construct a device instance and attach it to this instance on host
+  TrackManager<Track> *ConstructOnDevice()
+  {
+#ifndef COPCORE_CUDA_COMPILER
+    return nullptr; // should never happen
+#else
+    const size_t QueueSize  = adept::MParray::SizeOfInstance(fCapacity);
+    const size_t TracksSize = sizeof(Track) * fCapacity;
+    COPCORE_CUDA_CHECK(cudaMalloc(&fInstance_d, sizeof(TrackManager<Track>)));
+    COPCORE_CUDA_CHECK(cudaMalloc(&fActiveTracks, QueueSize));
+    COPCORE_CUDA_CHECK(cudaMalloc(&fNextTracks, QueueSize));
+    COPCORE_CUDA_CHECK(cudaMalloc(&fBuffer, TracksSize));
+
+    device_impl_trackmgr::construct_trackmanager<Track>
+        <<<1, 1>>>(fInstance_d, fCapacity, fActiveTracks, fNextTracks, fBuffer);
+    return fInstance_d;
+#endif
+  }
+
+  void FreeFromDevice()
+  {
+#ifdef COPCORE_CUDA_COMPILER
+    COPCORE_CUDA_CHECK(cudaFree(fBuffer));
+    COPCORE_CUDA_CHECK(cudaFree(fActiveTracks));
+    COPCORE_CUDA_CHECK(cudaFree(fNextTracks));
+    COPCORE_CUDA_CHECK(cudaFree(fInstance_d));
+#endif
+  }
+
+  /// @brief Swap active and next track slots. Compact if the fill percentage is higher than the threshold.
+  /// @details Must be called after the stats were updated on host.
+  template <typename Stream>
+  bool SwapAndCompact(float compact_threshold, Stream stream)
+  {
+#ifndef COPCORE_CUDA_COMPILER
+    COPCORE_EXCEPTION("SwapAndCompact cannot be called in hos-only code");
+    return false;
+#else
+    // check if the compacting threshold is hit
+    int used     = fStats.GetNused();
+    int inFlight = fStats.fInFlight;
+    assert(used >= 0 && used < fCapacity);
+    // Estimate maximum space needed if we DON'T compress now
+    int needed = used + 2 * inFlight;
+    if (needed < compact_threshold * fCapacity || needed > fCapacity) {
+      device_impl_trackmgr::swap_active<Track><<<1, 1, 0, stream>>>(fInstance_d);
+      if (needed > fCapacity) {
+        std::cout << "TrackManager::SwapAndCompact  ALERT: not enough space left to compress " << used
+                  << " used slots. Consider increasing TrackManager capacity.\n";
+      }
```

Comment:
Not sure this check is correct: `needed` is the space we need if we don't compress. For compressing, we actually only need an additional `inflight` slots, right?

Reply:
Yes, it is, `needed` takes into account also the space taken by the uncompressed tracks in the scenario where the compression will not be done.

Reply:
I understand, but we still only need `used + inFlight`, not twice `inFlight`.

Reply:
The worst-case scenario is that each particle `inFlight` will produce 2 extra particles (e.g. annihilation). This would require in total `2 * inFlight` extra slots. So to survive the next iteration conservatively, the buffer must have at least: `used + 2 * inflight` slots (because the current `inFlight` slots are already embedded in the `used` ones)

Reply:
Indeed I revisited the "not enough space to compress" condition, should be OK in the last version